rust: Add a Rust compatible API implementation.

soburi opened the pull request:
We will gradually add a Rust compatible API implementation conforming to Apache-2.
This PR adds an alternative implementation of Common.cpp.

soburi committed:
rust: Add a Rust compatible API implementation.

We will gradually add a Rust compatible API implementation conforming
to Apache-2.
This PR adds an alternative implementation of Common.cpp.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
CI: Prepare Rust modules

Adding Rust architecture modules for ensure compilation.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

pillo79 commented on Kconfig:
```diff
@@ -13,6 +13,7 @@ config ARDUINO_API
 	imply CBPRINTF_FP_SUPPORT
 	imply RING_BUFFER
 	select UART_INTERRUPT_DRIVEN
+	select RUST
```
I think this should currently be an optional feature, to avoid adding the strong dependency right away.

pillo79 reviewed (commented):
Awesome, looking forward to see this grow! 🥳 🤗

DhruvaG2000 commented on .github/workflows/build.yml:
```diff
@@ -39,6 +39,12 @@ jobs:
         run: |
           git clone https://github.com/arduino/ArduinoCore-API.git $MODULE_PATH/../ArduinoCore-API
           cp -rfp $MODULE_PATH/../ArduinoCore-API/api $MODULE_PATH/cores/arduino/
+          apt install rustup
```
The recommended way to install Rust is via the rustup.rs installer script, Ubuntu's rustup package from apt is often outdated and may not work correctly

DhruvaG2000 replied:
Also,
All initialization is lumped into one "Initialize" step mixing unrelated concerns (cloning API, copying files, Rust setup)...
  Better structure:
```
  - name: Clone ArduinoCore-API
[...]
 - name: Install Rust toolchain
[...]
```

soburi replied:
Fixed. Thank you for your comment.

DhruvaG2000 replied:
@soburi  you didn't address the first comment, why are we not using `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh` instead of apt?

soburi replied:
I missed it. It was addressed.

DhruvaG2000 commented on .github/workflows/build.yml:
```diff
@@ -39,6 +39,12 @@ jobs:
+          rustup target add thumbv6m-none-eabi
+          rustup target add thumbv7em-none-eabihf
+          rustup target add thumbv7em-none-eabi
+          rustup target add thumbv7m-none-eabi
```
Installing targets one-by-one is slower. Rustup supports installing multiple targets in one command

DhruvaG2000 commented on .github/workflows/build.yml:
```diff
@@ -39,6 +39,12 @@ jobs:
+          rustup target add thumbv6m-none-eabi
+          rustup target add thumbv7em-none-eabihf
+          rustup target add thumbv7em-none-eabi
+          rustup target add thumbv7m-none-eabi
```
```suggestion
          rustup target add thumbv6m-none-eabi thumbv7em-none-eabihf thumbv7em-none-eabi thumbv7m-none-eabi
```

soburi replied:
Fixed.

DhruvaG2000 commented on rust/src/common.rs:
```diff
@@ -0,0 +1,22 @@
+    let num = x.wrapping_sub(in_min).wrapping_mul(out_max.wrapping_sub(out_min));
+    let den = in_max.wrapping_sub(in_min);
+    // Note: To keep compatibility, the panic when den=0 is left as is.
+    num / den.wrapping_add(out_min)
```
Due to operator precedence, this evaluates as:             
```                                                                                                                                                      
  num / (den + out_min)  // Division happens AFTER adding out_min to den!                                                                                                                                          
```
Original C++ code:
```cpp
  return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
```

Can we just follow similar arithmetic syntax in rust too, the code using `wrapping_*` is making the logic really confusing to understand, and with rust already being a new addition I feel like we should make an effort to keep things simple where possible.

soburi replied:
Since this is not a "new feature" but a reimplementation of the original implementation, we believe compatibility should be maintained.

From another perspective, this implementation is an internal implementation of the C++ API,
and therefore does not take advantage of Rust's features.
To take advantage of Rust's features would require a complete overhaul of the API set, so we think it's better to leave it this way for now.
It would also be good to have an option for more Rust-like behavior in the future.

DhruvaG2000 replied:
ok it seems like there were some debates on how the behaviour shoul look like on the original Arduino API:
https://github.com/arduino/Arduino/issues/10382
and
https://github.com/arduino/ArduinoCore-API/issues/51

I will have to read up a bit to understand what's going on with map() in arduino

DhruvaG2000 replied:
Will get back to this at a later point, anyway no objections for this PR

pillo79 commented on Kconfig:
```diff
@@ -17,6 +17,11 @@ config ARDUINO_API
 
+config USE_ARDUINO_API_RUST_IMPLEMENTATION
+	bool "Use Rust implementation API core"
+	default y
```
Sorry to be nitpicky, but I asked this before and I'm still confused. 🙂 

Is Rust expected to be installed by anyone trying to use the core at this stage?
If not, I would expect this to be manually set somehow (via the west command line, using snippets, etc) and not being `default y`-ed.

DhruvaG2000 replied:
+1 @pillo79 
I would prefer this be optional and not default y too.

soburi replied:
OK. This is still in an experimental state.
Removed the `default y`.

pillo79 commented on Kconfig:
```diff
@@ -17,6 +17,10 @@ config ARDUINO_API
 
 if ARDUINO_API
 
+config USE_ARDUINO_API_RUST_IMPLEMENTATION
```
I am missing where this symbol is effectively used? AFAICS  CMakeLists.txt will try to invoke `cargo build` regardless of this being set or not.

soburi replied:
Oops, I missed that.
Adding if () ...

pillo79 reviewed (commented):
Sorry, I might be missing something huge, but... it still seems to me that effectively Rust is being forced on every core use. What is gating the new `add_custom_command` in the main CMakeLists.txt?

Again, I think it's awesome to see this working 😍, but IMO it should be optional at this stage.

soburi committed:
Arduino.h: Give priority to include path for the command line.

ArduinoAPI.h is loaded with priority from the path specified
on the command line.
This is part of coexisting with a rust implementation.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>
